{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b6b6e165",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FUNCTION IMPLEMENTATIONS\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# dataframe of DNA to RNA base pairings\n",
    "rna_dict = pd.DataFrame({'dna_nuc': ['A', 'T', 'C', 'G'], \n",
    "                         'rna_nuc': ['U', 'A', 'G', 'C']})\n",
    "rna_dict = rna_dict.set_index('dna_nuc')\n",
    "\n",
    "# returns mRNA based on coding strand DNA \n",
    "def dna_to_rna(dna):\n",
    "    mrna = np.array([])\n",
    "    \n",
    "    for i in range(dna.size):\n",
    "        bp = dna[0][i]\n",
    "        \n",
    "        if bp != 'N':\n",
    "            rna_nuc = rna_dict['rna_nuc'][bp]\n",
    "            mrna = np.append(mrna, rna_nuc)\n",
    "        \n",
    "    return np.expand_dims(mrna, axis=0)       \n",
    "\n",
    "# dataframe of mRNA codon to amino acid pairings\n",
    "aa_dict = pd.read_excel('amino_acid_dict.xlsx')\n",
    "aa_dict = aa_dict.set_index('codon')\n",
    "\n",
    "# returns the corresponding polypeptide chain to mRNA \n",
    "def rna_to_polypeptide(mrna): \n",
    "    polypeptide = np.array([])\n",
    "    \n",
    "    total_codons = mrna.size // 3\n",
    "    base_sequence = mrna[0][:(total_codons * 3)]\n",
    "#     print(mrna.size)\n",
    "#     print(base_sequence.size)\n",
    "#     print(total_codons * 3)\n",
    "    codon_sequence = base_sequence.reshape((total_codons, 3))\n",
    "    \n",
    "    for i in range(total_codons):\n",
    "        codon = codon_sequence[i]\n",
    "        codon = str(codon[0]) + str(codon[1]) + str(codon[2])\n",
    "        peptide = aa_dict['amino acid'][codon]\n",
    "        polypeptide = np.append(polypeptide, peptide)\n",
    "    \n",
    "    return np.expand_dims(polypeptide, axis=0)\n",
    "\n",
    "# sum total number of each key phrase in keys found in sequence\n",
    "def total_instances_of(sequence, keys):\n",
    "    count = 0\n",
    "    \n",
    "    for i in range(len(keys)):\n",
    "        count += np.sum(np.char.count(sequence, keys[i]))\n",
    "        \n",
    "    return count\n",
    "\n",
    "# test cases\n",
    "# dna = np.expand_dims(np.array(list('AAAGGGCCCTTT')), axis=0)\n",
    "# mrna = dna_to_rna(dna)\n",
    "# polypeptide = rna_to_polypeptide(mrna)\n",
    "\n",
    "# print(mrna)\n",
    "# print(polypeptide)\n",
    "# print(total_instances_of(polypeptide, ['phe', 'pro']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "1265a2f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num of gene sequences: 4380\n",
      "(4380, 25)\n"
     ]
    }
   ],
   "source": [
    "# FEATURE EXTRACTION\n",
    "\n",
    "genomic_data = pd.read_table('human_data.txt') # columnns: 'sequence', 'class'\n",
    "genomic_data.head()\n",
    "num_sequences = genomic_data.shape[0]\n",
    "print(\"Num of gene sequences: \" + str(num_sequences))\n",
    "\n",
    "training_data = np.empty((0, 25))\n",
    "\n",
    "# convert all gene training examples into feature vectors\n",
    "for i in range(num_sequences):\n",
    "    dna = np.expand_dims(np.array(list(human_data['sequence'][i])), axis=0)\n",
    "    mrna = dna_to_rna(dna)\n",
    "    polypeptide = rna_to_polypeptide(mrna)\n",
    "\n",
    "    non_polar_aa = ['gly', 'ala', 'val', 'cys', 'pro', 'leu', 'ile', 'met', 'trp', 'phe']\n",
    "    polar_aa = ['ser', 'thr', 'tyr', 'asn', 'gln']\n",
    "    pos_aa = ['lys', 'arg', 'his']\n",
    "    neg_aa = ['asp', 'glu']\n",
    "\n",
    "    total_bp = total_instances_of(dna, ['A', 'T', 'C', 'G'])\n",
    "    total_A = total_instances_of(dna, ['A'])\n",
    "    total_T = total_instances_of(dna, ['T'])\n",
    "    total_C = total_instances_of(dna, ['C'])\n",
    "    total_G = total_instances_of(dna, ['G'])\n",
    "    total_gly = total_instances_of(polypeptide, ['gly'])\n",
    "    total_ala = total_instances_of(polypeptide, ['ala'])\n",
    "    total_val = total_instances_of(polypeptide, ['val'])\n",
    "    total_cys = total_instances_of(polypeptide, ['cys'])\n",
    "    total_pro = total_instances_of(polypeptide, ['pro'])\n",
    "    total_leu = total_instances_of(polypeptide, ['leu'])\n",
    "    total_ile = total_instances_of(polypeptide, ['ile'])\n",
    "    total_met = total_instances_of(polypeptide, ['met'])\n",
    "    total_trp = total_instances_of(polypeptide, ['trp'])\n",
    "    total_phe = total_instances_of(polypeptide, ['phe'])\n",
    "    total_ser = total_instances_of(polypeptide, ['ser'])\n",
    "    total_thr = total_instances_of(polypeptide, ['thr'])\n",
    "    total_tyr = total_instances_of(polypeptide, ['tyr'])\n",
    "    total_asn = total_instances_of(polypeptide, ['asn'])\n",
    "    total_gln = total_instances_of(polypeptide, ['gln'])\n",
    "    total_lys = total_instances_of(polypeptide, ['lys'])\n",
    "    total_arg = total_instances_of(polypeptide, ['arg'])\n",
    "    total_his = total_instances_of(polypeptide, ['his'])\n",
    "    total_asp = total_instances_of(polypeptide, ['asp'])\n",
    "    total_glu = total_instances_of(polypeptide, ['glu'])\n",
    "    \n",
    "#     total_non_polar = total_instances_of(polypeptide, non_polar_aa)\n",
    "#     total_polar = total_instances_of(polypeptide, polar_aa)\n",
    "#     total_pos = total_instances_of(polypeptide, pos_aa)\n",
    "#     total_neg = total_instances_of(polypeptide, neg_aa)\n",
    "    \n",
    "#     feature_vector1 = np.array([total_A, total_T, total_C, total_G]) #/ total_bp\n",
    "#     feature_vector2 = np.array([total_non_polar, total_polar, total_pos, total_neg]) #/ (total_bp / 3)\n",
    "#     feature_vector = np.append(feature_vector1, feature_vector2, axis=0)\n",
    "\n",
    "    feature_vector = np.array([total_bp, total_A, total_T, total_C, total_G, total_gly, total_ala, total_val, total_cys, total_pro, total_leu,\n",
    "                            total_ile, total_met, total_trp, total_phe, total_ser, total_thr, total_tyr, total_asn, total_gln, total_lys,\n",
    "                            total_arg, total_his, total_asp, total_glu]) \n",
    "    \n",
    "    feature_vector = np.expand_dims(feature_vector, axis=0)\n",
    "    # print(feature_vector.shape)\n",
    "    training_data = np.append(training_data, feature_vector, axis=0)\n",
    "    # print('iteration ' + str(i))\n",
    "\n",
    "print(training_data.shape)\n",
    "\n",
    "# save training data to Excel file\n",
    "labels = np.expand_dims(genomic_data['class'], axis=1)\n",
    "training_data = np.append(training_data, labels, axis=1)\n",
    "\n",
    "df = pd.DataFrame(training_data)\n",
    "df.to_excel(excel_writer = r'C:\\Users\\Nathan\\Deep Learning\\genome-to-protein-classifier\\gene_features_dataset.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "dce54440",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_excel(excel_writer = r'C:\\Users\\Nathan\\Deep Learning\\genome-to-protein-classifier\\gene_features_dataset.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4dfd2e8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['tyr' 'gly' 'val' 'asp' 'leu' 'stop' 'trp' 'his' 'thr' 'gly' 'trp'\n",
      "  'tyr' 'stop' 'trp' 'gly' 'tyr' 'glu' 'glu' 'cys' 'asp' 'lys' 'glu'\n",
      "  'stop' 'trp' 'val' 'asp' 'phe' 'tyr' 'asn' 'leu' 'cys' 'leu' 'met'\n",
      "  'val' 'asp' 'gly' 'gly' 'ser' 'gly' 'phe' 'gly' 'tyr' 'phe' 'tyr' 'phe'\n",
      "  'leu' 'ile' 'leu' 'phe' 'gly' 'thr' 'leu' 'gly' 'phe' 'thr' 'cys' 'phe'\n",
      "  'stop' 'thr' 'ser' 'glu' 'val' 'ser' 'asn' 'gly' 'gly' 'val' 'arg'\n",
      "  'ile']]\n"
     ]
    }
   ],
   "source": [
    "genomic_data = pd.read_table('human_data.txt') # columnns: 'sequence', 'class'\n",
    "dna = np.expand_dims(np.array(list(genomic_data['sequence'][0])), axis=0)\n",
    "\n",
    "result = rna_to_polypeptide(dna_to_rna(dna))\n",
    "result = np.reshape(result, (1, result.shape[0] * result.shape[1]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a392946a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8 (tensorflow)",
   "language": "python",
   "name": "tensorflow"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
